### ISSUE TITLE
L1 memory error in ResNet50_Backbone_with_FPN_V2

### ISSUE THREAD

**Issue by kamalrajkannan78** [tenstorrent/tt-xla#3389]
# RetinaNet L1 memory error

### Describe the bug

RetinaNet (ResNet50 Backbone with FPN V2) inference fails on TT with an L1 memory error when moving model output to host. The TT runtime reports:

**Full model (whole RetinaNet):**
```
Statically allocated circular buffers on core range [(x=0,y=0) - (x=7,y=6)] grow to 22734112 B which is beyond max L1 size of 1499136 B
```

The failure occurs in **UntilizeWithUnpadding** during `toHost(untilize=true)` when the test (or any consumer) moves the detection tensors from device to CPU.

---

### Root cause analysis

The error originates in **torchvision’s RetinaNet** postprocessing path, not in the backbone or FPN. We sliced the model block-by-block and isolated the failure to the **boolean mask indexing** (masked_select) in `postprocess_detections`.

**Where the error comes from in the model (torchvision):**

| Location | File | Line | What |
|----------|------|------|------|
| Entry | `torchvision/models/detection/retinanet.py` | 667 | `detections = self.postprocess_detections(...)` |
| **Exact line (L1 trigger)** | `torchvision/models/detection/retinanet.py` | **551** | `scores_per_level = scores_per_level[keep_idxs]` |

That line is **boolean mask indexing**: it keeps only the scores above threshold (`keep_idxs = scores_per_level > self.score_thresh`). In PyTorch/ATen this is **`aten::masked_select`**. The result (and/or downstream tensors built from it) stays on device in tiled layout; when the harness **moves the final detections to host**, the runtime runs **UntilizeWithUnpadding** to convert tiled → row-major. For this graph, the untilize buffer size exceeds L1 (22.7 MB > 1.5 MB), so the TT runtime throws.

**Code path to the problematic op:**

```
RetinaNet.forward()
└── postprocess_detections(head_outputs, anchors, image_shapes)   [retinanet.py ~509]
    └── per image, per FPN level:
        ├── scores_per_level = torch.sigmoid(logits_per_level).flatten()
        ├── keep_idxs = scores_per_level > self.score_thresh
        ├── scores_per_level = scores_per_level[keep_idxs]   ← L1 TRIGGER (masked_select)
        ├── topk_idxs = torch.where(keep_idxs)[0]
        ├── topk, decode_single, clip_boxes_to_image, append...
    └── torch.cat(...), batched_nms, detections.append({ boxes, scores, labels })
→ move_to_host(detections) → toHost(untilize=true) → UntilizeWithUnpadding → L1 overflow
```

**Exact place of the problematic layer:**

`retinanet.py:551` → `scores_per_level[keep_idxs]` (boolean indexing) → `aten::masked_select` → downstream untilize on move-to-host → **UntilizeWithUnpadding** (TT) → L1 buffer 22734112 B > 1499136 B.

---

### L1 error replication: sanity model vs full model

The **same class of L1 error** (UntilizeWithUnpadding buffer beyond L1 limit) is **replicated in a minimal sanity model** that only does the same op (flatten → threshold → `tensor[mask]`), but the **reported buffer size and core range differ**:

| Case | Buffer size | Core range | Notes |
|------|-------------|------------|--------|
| **Full model (RetinaNet)** | **22734112 B** | (x=0,y=0) - (x=7,y=6) | Whole RetinaNet inference; move-to-host triggers L1. |
| **Sanity model** (masked_select only) | **22803744 B** | (x=7,y=6) - (x=7,y=6) | Same op (masked_select), same failure type; values not matching. |

So: **the L1 memory failure is replicated** (same op, same UntilizeWithUnpadding overflow), but **the exact buffer size and core range are not the same** between the full model and the sanity model. The sanity test uses a fixed input size (5683528 elements, from 22734112/4); the full model has different tensor shapes and graph structure, so the allocated buffer and core range differ.

---

### Cases

**Case 1 – Full RetinaNet model (L1 error)**  
- Run: `pytest tests/runner/test_models.py::test_all_models_torch[retinanet/pytorch-ResNet50_Backbone_with_FPN_V2-single_device-inference] -v -s`  
- Result: TT_THROW – circular buffers 22734112 B beyond max L1 1499136 B on core range (x=0,y=0)-(x=7,y=6).  

I have attached the log for your reference:[retinanet-ResNet50_Backbone_with_FPN_V2-single_device-inference.log](https://github.com/user-attachments/files/25569899/retinanet-ResNet50_Backbone_with_FPN_V2-single_device-inference.log)

**Case 2 – Sanity model (masked_select only, L1 replicated)**  
- Test: `tests/torch/graphs/test_untilize_with_unpadding_l1_sanity.py::test_masked_select_l1_sanity`  
- Model: flatten → `x[x > thresh]` (same op as retinanet.py:551).  
- Result: L1 error replicated (UntilizeWithUnpadding), but **22803744 B** on core **(x=7,y=6)-(x=7,y=6)** (values differ from full model).  

python code with sanity:[test_untilize_with_unpadding_l1_sanity.py](https://github.com/user-attachments/files/25569902/test_untilize_with_unpadding_l1_sanity.py)
I have attached the log file for your reference:[sanity_l1.log](https://github.com/user-attachments/files/25569920/sanity_l1.log)

**Case 3 – sanity with one more layer**  
- When try adding one more layer to sanity 
logs for your reference:[retinanet_one_line_before_error.log](https://github.com/user-attachments/files/25569947/retinanet_one_line_before_error.log)

---

### Conclusion

- **Root cause:** The L1 error is triggered when moving RetinaNet’s detection output to host. The TT runtime runs **UntilizeWithUnpadding** for that transfer; the buffer size exceeds L1. The **code path** that produces the large tensor includes **`retinanet.py:551`** – `scores_per_level = scores_per_level[keep_idxs]` (**masked_select**).
- **Replication:** The same **type** of L1 error is **replicated** in a minimal sanity test that only runs the same op (masked_select). The **exact buffer size and core range** in the error message **do not match** between full model (22734112 B, core range (x=0,y=0)-(x=7,y=6)) and sanity model (22803744 B, core (x=7,y=6)-(x=7,y=6)); both are the same **UntilizeWithUnpadding L1 overflow**.
- **Exact location in torchvision:** **`torchvision/models/detection/retinanet.py` line 551** – boolean mask indexing in `postprocess_detections`, lowered to **aten::masked_select**, with UntilizeWithUnpadding failing on move-to-host.

**Comment by kamalrajkannan78** [comment]
fixed via https://github.com/tenstorrent/tt-forge-models/pull/533
